kitsune/wiki/views.py

```diff
@@ -452,125 +452,154 @@ def steal_lock(request, document_slug, revision_id=None):
     return HttpResponse("", status=200 if ok else 400)
 
 
-@require_http_methods(["GET", "POST"])
-@login_required
-def edit_document(request, document_slug, revision_id=None):
-    """Create a new revision of a wiki document, or edit document metadata."""
+def _get_user_and_doc(request, document_slug):
     user = request.user
-
     doc = get_visible_document_or_404(
         user,
         locale=request.LANGUAGE_CODE,
         slug=document_slug,
         look_for_translation_via_parent=True,
         return_parent_if_no_translation=True,
     )
+    return user, doc
+
+
+def _check_edit_perms(user, doc, doctype):
+    if doctype == "doc":
+        if not doc.allows(user, "create_revision"):
+            raise PermissionDenied
+    if doctype == "meta":
+        if not doc.allows(user, "edit") or not user.is_staff:
+            raise PermissionDenied
+
+
+def _get_rev(doc, revision_id):
+    if revision_id:
+        rev = get_object_or_404(Revision, pk=revision_id, document=doc)
+    else:
+        rev = doc.current_revision or doc.revisions.order_by("-created", "-id")[0]
+
+    return rev
+
+
+@require_http_methods(["GET", "POST"])
+@login_required
+def edit_document(request, document_slug, revision_id=None):
```

Comment:
`edit_document` and `edit_document_metadata` share a lot of code. We could use something like 

```python
def base_edit_checks(request, document_slug, revision_id=None, doctype="doc"):

    user = request.user
    doc = get_visible_document_or_404(
        user,
        locale=request.LANGUAGE_CODE,
        slug=document_slug,
        look_for_translation_via_parent=True,
        return_parent_if_no_translation=True,
    )

    if doctype == "doc":
        if not doc.allows(user, "create_revision"):
            raise PermissionDenied
    if doctype == "meta":
        if not doc.allows(user, "edit") or not user.is_staff:
            raise PermissionDenied

    if doc.locale != request.LANGUAGE_CODE:
        # We've fallen back to the parent, since no visible translation existed.
        url = reverse("wiki.translate", locale=request.LANGUAGE_CODE, args=[document_slug])
        return HttpResponseRedirect(url)

    # If this document has a parent, then the edit is handled by the
    # translate view. Pass it on.
    if doc.parent:
        return translate(request, doc.parent.slug, revision_id)

    if revision_id:
        rev = get_object_or_404(Revision, pk=revision_id, document=doc)
    else:
        rev = doc.current_revision or doc.revisions.order_by("-created", "-id")[0]

    return user, doc, rev
```
and remove all the 3 helper functions used that start with the underscore. The `edit_document_metadata` view needs to add a forth param like `user, doc = base_edit_checks(request, document_slug, revision_id=None, doctype="meta")`

Reply:
The above will remove a lot of duplicate code and each view can focus on its respective function. With that said, I believe it's worth adding an issue to rethink how the forms are structured. I believe we can remove a lot of boilerplate code. That said, it's working fine and it's not a priority

Reply:
The locale piece that returns an `HttpResponseRedirect` and the doc.parent check that returns `translate` both have to be returned again in the main view function. So, for instance, when I tried to break the locale piece into it's own function, I still had to have this in the view:
```
if (ret := _check_locale(request, document_slug, doc) is not None:
    return ret
```

Which seemed less readable than just leaving the code in place.  Thoughts?

Reply:
I'll implement - thanks.